RViz crash with high-resolution map

RViz seems to have trouble rendering high-pixel count maps, causing a very reproducible crash.
Publishing the map in `map_lores.zip` (533x423 px, 230.3 kB) and viewing it in rviz2 works fine.
Publishing the map in `map_hires.zip` (861x661 px, 569.2 kB) and trying to view it in rviz2 causes a hard crash with the following error info:

```
ros2 launch openrover_demo rviz.launch.py 
[INFO] [launch]: All log files can be found below /home/snuc/.ros/log/2019-07-05-16-22-21-496956-rhea-15882
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [rviz2-1]: process started with pid [15913]
[rviz2-1] [INFO] [rviz2]: Stereo is NOT SUPPORTED
[rviz2-1] [INFO] [rviz2]: OpenGl version: 3 (GLSL 1.3)
[rviz2-1] [INFO] [rviz2]: Stereo is NOT SUPPORTED
[rviz2-1] [INFO] [rviz2]: Creating 1 swatches_
[rviz2-1] [ERROR] [rviz2]: InvalidParametersException: Stream size does not match calculated image size in Image::loadRawData at /opt/ros/master/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/OgreMain/src/OgreImage.cpp (line 283)
[rviz2-1] [ERROR] [rviz2]: Failed to create 1 swatches_
[rviz2-1] [INFO] [rviz2]: Creating 2 swatches_
[rviz2-1] [ERROR] [rviz2]: InvalidParametersException: Stream size does not match calculated image size in Image::loadRawData at /opt/ros/master/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/OgreMain/src/OgreImage.cpp (line 283)
[rviz2-1] [ERROR] [rviz2]: Failed to create 2 swatches_
[rviz2-1] [INFO] [rviz2]: Creating 4 swatches_
[rviz2-1] [ERROR] [rviz2]: InvalidParametersException: Stream size does not match calculated image size in Image::loadRawData at /opt/ros/master/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/OgreMain/src/OgreImage.cpp (line 283)
[rviz2-1] [ERROR] [rviz2]: Failed to create 4 swatches_
[rviz2-1] [INFO] [rviz2]: Creating 8 swatches_
[rviz2-1] [ERROR] [rviz2]: InvalidParametersException: Stream size does not match calculated image size in Image::loadRawData at /opt/ros/master/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/OgreMain/src/OgreImage.cpp (line 283)
[rviz2-1] [ERROR] [rviz2]: Failed to create 8 swatches_
[rviz2-1] [ERROR] [rviz2]: InvalidParametersException: Stream size does not match calculated image size in Image::loadRawData at /opt/ros/master/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/OgreMain/src/OgreImage.cpp (line 283)
[rviz2-1] terminate called after throwing an instance of 'Ogre::InvalidParametersException'
[rviz2-1]   what():  InvalidParametersException: Stream size does not match calculated image size in Image::loadRawData at /opt/ros/master/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/OgreMain/src/OgreImage.cpp (line 283)
[ERROR] [rviz2-1]: process has died [pid 15913, exit code -6, cmd '/opt/ros/master/install/lib/rviz2/rviz2 --display-config /home/snuc/ros2_ws/src/openrover/openrover_demo/config/default.rviz --fixed-frame map'].
```

This is on Ubuntu 18 with ROS2 built from master branch.

[map_hires.zip](https://github.com/ros2/rviz/files/3364037/map_hires.zip)
[map_lores.zip](https://github.com/ros2/rviz/files/3364040/map_lores.zip)

--- a/rviz_default_plugins/src/rviz_default_plugins/displays/map/map_display.cpp
+++ b/rviz_default_plugins/src/rviz_default_plugins/displays/map/map_display.cpp
@@ -334,7 +334,8 @@ void MapDisplay::createSwatches()
   const size_t maximum_number_swatch_splittings = 4;
 
   for (size_t i = 0; i < maximum_number_swatch_splittings; ++i) {
-    RVIZ_COMMON_LOG_INFO_STREAM("Creating " << number_swatches << " swatches_");
+    RVIZ_COMMON_LOG_INFO_STREAM("Trying to create a map of size " <<
+      width << " x " << height << " using " << number_swatches << " swatches");
     swatches_.clear();
     try {
       tryCreateSwatches(width, height, resolution, swatch_width, swatch_height, number_swatches);
@@ -355,7 +356,8 @@ void MapDisplay::createSwatches()
 void MapDisplay::doubleSwatchNumber(
   size_t & swatch_width, size_t & swatch_height, int & number_swatches) const
 {
-  RVIZ_COMMON_LOG_ERROR_STREAM("Failed to create " << number_swatches << " swatches_");
+  RVIZ_COMMON_LOG_ERROR_STREAM("Failed to create map using " << number_swatches << " swatches. "
+    "At least one swatch seems to need too much memory");
   if (swatch_width > swatch_height) {
     swatch_width /= 2;
   } else {

--- a/rviz_default_plugins/src/rviz_default_plugins/displays/map/map_display.cpp
+++ b/rviz_default_plugins/src/rviz_default_plugins/displays/map/map_display.cpp
@@ -328,13 +328,17 @@ void MapDisplay::createSwatches()
   size_t swatch_width = width;
   size_t swatch_height = height;
   int number_swatches = 1;
-  int maximum_number_swatch_splittings = 4;  // 4 seems to work well for this purpose.
+  // One swatch can have up to 2^16 * 2^16 pixel (8 bit texture, i.e. 4GB of data)
+  // Since the width and height are separately limited by 2^16 it might be necessary to have several
+  // pieces, however more than 8 swatches is probably unnecessary due to memory limitations
+  const size_t maximum_number_swatch_splittings = 4;
 
-  for (int i = 0; i < maximum_number_swatch_splittings; i++) {
+  for (size_t i = 0; i < maximum_number_swatch_splittings; ++i) {
     RVIZ_COMMON_LOG_INFO_STREAM("Creating " << number_swatches << " swatches_");
     swatches_.clear();
     try {
       tryCreateSwatches(width, height, resolution, swatch_width, swatch_height, number_swatches);
+      updateDrawUnder();
       return;
     } catch (Ogre::InvalidParametersException &) {
       doubleSwatchNumber(swatch_width, swatch_height, number_swatches);
@@ -343,7 +347,9 @@ void MapDisplay::createSwatches()
       doubleSwatchNumber(swatch_width, swatch_height, number_swatches);
     }
   }
-  updateDrawUnder();
+  RVIZ_COMMON_LOG_ERROR_STREAM("Creating " << number_swatches << "failed. "
+    "This map is too large to be displayed by RViz.");
+  swatches_.clear();
 }
 
 void MapDisplay::doubleSwatchNumber(

--- a/rviz_default_plugins/src/rviz_default_plugins/displays/map/swatch.cpp
+++ b/rviz_default_plugins/src/rviz_default_plugins/displays/map/swatch.cpp
@@ -209,7 +209,7 @@ void Swatch::resetTexture(Ogre::DataStreamPtr & pixel_stream)
     "MapTexture" + std::to_string(texture_count_++),
     "rviz_rendering",
     pixel_stream,
-    static_cast<uint8_t>(width_), static_cast<uint8_t>(height_),
+    static_cast<uint16_t>(width_), static_cast<uint16_t>(height_),
     Ogre::PF_L8, Ogre::TEX_TYPE_2D, 0);
 }
 

--- a/rviz_default_plugins/test/rviz_default_plugins/displays/map/map_display_test.cpp
+++ b/rviz_default_plugins/test/rviz_default_plugins/displays/map/map_display_test.cpp
@@ -204,10 +204,11 @@ TEST_F(MapTestFixture, reset_deletes_map) {
 }
 
 TEST_F(MapTestFixture, createSwatches_creates_more_swatches_if_map_is_too_big) {
-  map_display_->processMessage(createMapMessage(500, 500));
+  // one dimension is larger than 2^16 --> that's too much for one texture buffer
+  map_display_->processMessage(createMapMessage(70000, 50));
 
   auto manual_objects = rviz_rendering::findAllOgreObjectByType<Ogre::ManualObject>(
     scene_manager_->getRootSceneNode(), "ManualObject");
 
-  EXPECT_THAT(manual_objects, SizeIs(4));
+  EXPECT_THAT(manual_objects, SizeIs(2));
 }

